### ISSUE TITLE
Multitenant Kafka integration with vmagent

### ISSUE THREAD

**Issue by Alon-Katz** [VictoriaMetrics/VictoriaMetrics#2212]
Hi, we are looking into the enterprise edition of VM, especially for the kafka integration with vmagent. 
But I cant figure out how does the write and read from kafka handles tenants.

vmagent supports remote write proxy, this means I can do remote write into vmagent with the tenant id and it would insert it into the correct tenant in VM cluster, but when I need the vmagent to first write the data into kafka (for multiple tenant ids) and then later on I want another vmagent to read the data from kafka and insert it into VM cluster using the right tenant id that this data is relevant to, how can this be setup?

I hope I am missing something simple here but it feels like if this is not supported, then its hardly an enterprise feature because without multi-tenancy support in this case it has little of value.

**Comment by f41gh7** [comment]
Hello, this configuration is supported with limitation - all vmagents must be started with flag - `-remoteWrite.multitenantURL`.

 For example, you have following setup: data-producer, vmagent-1, kafka, vmagent-2.
1)  data-producer writes multitenant data into vmagent-1. 
2) vmagent-1 has configuration for remote write `-remoteWrite.multitenantURL=kafka://kafka-init-broker/?topic=multitenant-rw`
3) vmagent writes data into kafka and adds special header to each message - `vm_tenant_id` with value of tenant_id
4) vmagent-2 has configration for remote write `-remoteWrite.multitenantURL=http://vminsert/` and `-kafka.consumer.topic=mutitenant-rw` `-kafka.consumer.topic.brokers=kafka-init-broker`

 You cannot use flag `-remoteWrite.URL` - vmagent will panic.

**Comment by f41gh7** [comment]
It must be reflected at doc, going to make an update for it.

**Comment by Alon-Katz** [comment]
Awesome! That's exactly what I was looking for, thanks!

**Comment by Alon-Katz** [comment]
Unfortunately this doesn't work, as long as the data-producer is only sending a single tenant that this setup works. But when the data-producer is sending more then 1 tenant, the vmagent is starting to throw errors and enters a crash loop.

This are the errors that it throws:
```
panic: BUG: metric "vmagent_remotewrite_kafka_queue_size{topic=\"metrics\",broker=\"broker-dns-removed-for-security:9092\"}" is already registered

goroutine 437 [running]:
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper.func1()
	github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:246 +0x65
panic({0xc51840, 0xc0000a21f0})
	runtime/panic.go:1038 +0x215
github.com/VictoriaMetrics/metrics.(*Set).mustRegisterLocked(0xc00014a1c0, {0xc0009c8070, 0xcbe780}, {0xde9ec0, 0xc0001f6978})
	github.com/VictoriaMetrics/metrics@v1.18.1/set.go:450 +0x219
github.com/VictoriaMetrics/metrics.(*Set).registerMetric(0xc00014a1c0, {0xc0009c8070, 0x6b}, {0xde9ec0, 0xc0001f6978})
	github.com/VictoriaMetrics/metrics@v1.18.1/set.go:433 +0xe5
github.com/VictoriaMetrics/metrics.(*Set).NewCounter(...)
	github.com/VictoriaMetrics/metrics@v1.18.1/set.go:126
github.com/VictoriaMetrics/metrics.NewCounter(...)
	github.com/VictoriaMetrics/metrics@v1.18.1/counter.go:20
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/kafka.newClient(0xc0000f01b0, {{0x7ffd81ae64f2, 0x10}, {0x7ffd81ae6491, 0x40}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}}, ...)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/kafka/producer_cgo.go:110 +0xc85
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/kafka.newKafkaClient(...)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/kafka/producer_cgo.go:21
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/kafka.NewClient(0xb1ed931a61184063, {{0x7ffd81ae64f2, 0x10}, {0x7ffd81ae6491, 0x40}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}}, ...)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/kafka/kafka.go:28 +0x38
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.newKafkaClient(0xc00014a1c0, 0xc000152880, 0xc0023dae78, {0xc00033e460, 0x10}, 0xc0001660e0)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/client.go:437 +0x2bb
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.newRemoteWriteCtx(0x0, 0x8, 0xc0000f01b0, 0x3, {0xc00033e460, 0x10})
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/remotewrite.go:446 +0x485
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.newRemoteWriteCtxs(0xc00033e420, {0xc0001117f0, 0x1, 0xc518a0})
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/remotewrite.go:199 +0x51e
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite.PushWithAuthToken(0xc0001521d0, 0xc000332190)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/remotewrite/remotewrite.go:264 +0x168
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite.insertRows(0xc00033e420, {0xc0003fe180, 0x6, 0x40f2e7}, {0x0, 0x0, 0x1})
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite/request_handler.go:84 +0x575
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite.InsertHandler.func1.1({0xc0003fe180, 0xc00038c000, 0x2ab})
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite/request_handler.go:34 +0x3a
github.com/VictoriaMetrics/VictoriaMetrics/lib/protoparser/promremotewrite.ParseStream({0x7f9f83d091d8, 0xc00037a2c0}, 0xc000a27558)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/protoparser/promremotewrite/streamparser.go:57 +0x602
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite.InsertHandler.func1()
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite/request_handler.go:33 +0x85
github.com/VictoriaMetrics/VictoriaMetrics/lib/writeconcurrencylimiter.Do(0xc002dea648)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/writeconcurrencylimiter/concurrencylimiter.go:37 +0x238
github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite.InsertHandler(0xc00033e420, 0xc000a52200)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/promremotewrite/request_handler.go:32 +0x91
main.processMultitenantRequest({0xdf0468, 0xc0000fc420}, 0xccd11e, {0xc0000ce065, 0xde1c10})
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/main.go:335 +0x8de
main.requestHandler({0xdf0468, 0xc0000fc420}, 0xc000a52200)
	github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent/main.go:312 +0x1652
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.handlerWrapper(0xc0000a23b0, {0xdf0468, 0xc0000fc420}, 0xc000a52200, 0xd6ee10)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:336 +0xc17
github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver.gzipHandler.func1({0xdf0738, 0xc0000fa0e0}, 0xc000110a60)
	github.com/VictoriaMetrics/VictoriaMetrics/lib/httpserver/httpserver.go:214 +0x76
net/http.HandlerFunc.ServeHTTP(0x0, {0xdf0738, 0xc0000fa0e0}, 0x40)
	net/http/server.go:2047 +0x2f
net/http.serverHandler.ServeHTTP({0xdefde0}, {0xdf0738, 0xc0000fa0e0}, 0xc000a52200)
	net/http/server.go:2879 +0x43b
net/http.(*conn).serve(0xc0001d03c0, {0xdf4700, 0xc0000fc2d0})
	net/http/server.go:1930 +0xb08
created by net/http.(*Server).Serve
	net/http/server.go:3034 +0x4e8
```

I will provide more information if needed.
Hopefully it can get fixed soon.

Thanks!

**Comment by f41gh7** [comment]
Get it, thanks for reporting, will be fixed soon.

**Comment by f41gh7** [comment]
Can you try following docker image: `victoriametrics/vmagent:v1.73.0-kafka-fix-enterprise` ?

 Please note, that kafka in this case used as transport, not a storage, even if remoteWrite for vmagent-2 will be unavailable, vmagent will fetch messages from kafka and store it at local disk buffer.

**Comment by Alon-Katz** [comment]
Hi @f41gh7, Thanks for the quick response, it works great.

**Comment by Alon-Katz** [comment]
Hi @f41gh7,
I just looked at the new version (1.73.1) released and I see that the issue that was fixed in victoriametrics/vmagent:v1.73.0-kafka-fix-enterprise still occurs, does the fix will be officaily released in the next version? if it will, do you know when will it be?

Thanks!
Hope you are all right in Kyiv! Stay strong! ❤️

**Comment by valyala** [comment]
@Alon-Katz , the bugfix has been included in [v1.74.0](https://docs.victoriametrics.com/CHANGELOG.html).

Note also that [vmagent](https://docs.victoriametrics.com/vmagent.html) and [VictoriaMetrics cluster](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html) gained improved support for multitenancy in [recent releases](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/latest) - see [these docs](https://docs.victoriametrics.com/vmagent.html#multitenancy).

Closing this feature request as done.
